### PR TITLE
[MIG] [11.0] project_closing: migration

### PR DESCRIPTION
**Pull request by Jerther**
Worked pretty well as is, with one difference. If you take a look in analytic.py, I had to add `active_test=False` in the context. Else, an inactive project would never reactivate. But that behavior wouldn't happen in the unit tests, and I couldn't replicate it either.

This leads me to think that this also happens in the 10.0 version if this module.

Any help to improve the tests to include the `active_test` problem is welcome.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please squash all "OCA Transbot.." commits following https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

**Comment by Jerther** [comment]
Anything else?

**Comment by pedrobaeza** [comment]
It seems no, but you need to find reviewers. People can spontaneously review it if they are interested, but if not, you can review other PRs from this repository or other similar, and ask in exchange that they review yours.

File: project_closing/README.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
```
Comment (emagdalenaC2i) [inline project_closing/README.rst]: Should be https://www.gnu.org/licenses/agpl as is defined in https://github.com/OCA/maintainer-tools/edit/master/template/module/README.rst ?

File: project_closing/README.rst [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Usage
+=====
+
+To use this module, you need to:
+
+* Go to *Project > Project > Projects*
+* Edit one
+* Click on *Archive*
+* The associated analytic account will also be archived.
+
```
Comment (emagdalenaC2i) [inline project_closing/README.rst:22]: Add this lines:
.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
   :alt: Try me on Runbot
   :target: https://runbot.odoo-community.org/runbot/project/11.0

File: project_closing/__init__.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
```
Comment (emagdalenaC2i) [inline project_closing/__init__.py]: Delete the utf-8 line and the copyright

File: project_closing/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
```
Comment (emagdalenaC2i) [inline project_closing/__manifest__.py]: Delete the utf-8 line and the copyright

File: project_closing/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    "name": "Project closing",
+    "version": "11.0.1.0.0",
+    "author": "Camptocamp, Métal Sartigan, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "website": "http://www.camptocamp.com",
+    "category": "Project Management",
```
Comment (emagdalenaC2i) [inline project_closing/__manifest__.py]: This should be "category": "Project" in order to agree with the other Odoo modules

File: project_closing/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    "name": "Project closing",
+    "version": "11.0.1.0.0",
+    "author": "Camptocamp, Métal Sartigan, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "website": "http://www.camptocamp.com",
```
Comment (emagdalenaC2i) [inline project_closing/__manifest__.py]: The website should be https://github.com/OCA/project 
See https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md

File: project_closing/models/analytic.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
```
Comment (emagdalenaC2i) [inline project_closing/models/analytic.py]: Remove the utf-8 line

File: project_closing/models/project.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
```
Comment (emagdalenaC2i) [inline project_closing/models/project.py]: Remove the utf-8 line

File: project_closing/tests/test_project.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
```
Comment (emagdalenaC2i) [inline project_closing/tests/test_project.py]: Remove the utf-8 line

**Review by emagdalenaC2i — APPROVED** [review]
LGTM (Technical and functional review)
Just a few comments about small changes

**Comment by Jerther** [comment]
@emagdalenaC2i  All done :)

File: project_closing/models/project.py [hunk 10]
```diff
@@ -9,8 +9,7 @@ class ProjectProject(models.Model):
     def toggle_active(self):
         if not self.env.context.get('doing_project_toggle_active'):
             # When called directly from Project, delegate to Analytic Account
-            res = self.analytic_account_id.toggle_active()
+            return self.analytic_account_id.toggle_active()
         else:
             # When called from the Analytic Account, perform the toggling
-            res = super(ProjectProject, self).toggle_active()
-        return res
+            return super(ProjectProject, self).toggle_active()
```
Comment (dreispt) [inline project_closing/models/project.py]: Sure its one LOC less, but I prefer to have a single exit point as much as possible. But then, its a subjective preference.

File: project_closing/README.rst (hunk 1 above)
Comment (dreispt) [reply project_closing/README.rst]: AGPL-3 is correct

File: project_closing/README.rst [hunk 11]
```diff
@@ -20,6 +20,10 @@ To use this module, you need to:
 * Click on *Archive*
 * The associated analytic account will also be archived.
 
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/{repo_id}/{branch}
```
Comment (dreispt) [inline project_closing/README.rst]: You forgot to replace the {repo_id} and {branch} placeholders

Comment (Jerther) [reply project_closing/README.rst]: Oh no... XD

File: project_closing/models/project.py (hunk 10 above)
Comment (Jerther) [reply project_closing/models/project.py]: True, and I mostly agree with that. Here's an excerpt from "Clean Code" which I like very much:

"Some programmers follow Edsger Dijkstra’s rules of structured programming. Dijkstra said that every function, and every block within a function, should have one entry and one exit. [...] those rules serve little benefit when functions are very small. [...] So if you keep your functions small, then the occasional multiple return, break, or continue statement does no harm and can sometimes even be more expressive than the single-entry, single-exit rule."

Subjective, true. I think this "exception" holds true mainly for strongly typed, nestable/narrow variable scope languages (C#, Java) where those principles make such a method a bit too verbose. Still, I found those 4 lines very small, and more expressive :)

In all cases, let's never do that in a 20+ lines method...

**Review by dreispt — APPROVED** [review]

**Review by emagdalenaC2i — APPROVED** [review]

**Comment by astirpe** [comment]
You may be interested in this: https://github.com/OCA/project/issues/353

File: project_closing/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{
+    "name": "Project closing",
+    "version": "11.0.1.0.0",
+    "author": "Camptocamp, Métal Sartigan, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "website": "https://github.com/OCA/project",
+    "category": "Project",
+    "depends": ["project"],
+    "data": [],
```
Comment (MeetKD) [inline project_closing/__manifest__.py:9]: Remove empty Key.

**Pull request merged**